Specify Captcha support

louischan-oursky opened the pull request:
Depend on https://github.com/authgear/authgear-server/pull/4399

pkong-ds commented on docs/specs/captcha.md:
````diff
@@ -1,41 +1,326 @@
-### authgear.yaml
+This section documents the old configuration.
 
-Before using any workflow with captcha enabled, the captcha provider must be configured in `authgear.yaml`:
+> The old configuration **IS NOT** used by Authentication Flow.
+> To configure Captcha providers for an Authentication Flow, the new configuration must be used.
 
-```yaml
+### authgear.yaml
+
+```
 captcha:
   provider: cloudflare
 ```
 
-- `provider`
-  - The captcha provider. Currently, only `cloudflare` is supported.
+- `captcha.provider`: The only possible value is `cloudflare`. The default is null.
 
 ### authgear.secrets.yaml
 
````
Suggestion - maybe move everything under `## Old configuration` into 
1. captcha_legacy.md; OR
2. move into [a collapsed section](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections#creating-a-collapsed-section)?

pkong-ds commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,328 @@
+
+### Default behavior
+
+When Captcha is enabled and there is at least one configured provider,
+all the branches of the first step (that is, the `identify` step) of the generated flows will have `captcha.required=true`.
+
+This means whether Captcha is enabled, every flow requires captcha at the beginning of the flow.
```
If 
- more than 1 configured provider
- user did not specify which provider to use

Will
1. throw error; OR
2. randomly select a configured provider; If this, any selection principle?

louischan-oursky replied:
Added a commit to clarify.

pkong-ds commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,328 @@
-### authgear.secrets.yaml
+Other fields are specific to `type`.
+
+#### `type: cloudflare`
```
Seems only cloudfare here, is recaptcha v2 intended to left blank for me to fill in when implementing? 🙏

louischan-oursky replied:
Added a commit to address this.

pkong-ds commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,354 @@
+- `items.type`: Required. It is the same as `captcha.providers.type`.
+- `items.alias`: Required. It is the same as `captcha.providers.alias`.
+
+Other fields are specific to `type`.
```
question: For implementation convenience, should we group these type-specific fields in a field call `configs` (open to suggestion)

Referenced how AuthenticationFlowObjectFlowRoot implemented `GetSteps()` for different Step variation of auth flow

```golang
type AuthenticationFlowObjectFlowRoot interface {
	AuthenticationFlowObject
	GetName() string
	GetSteps() []AuthenticationFlowObject
}
```

louischan-oursky replied:
Not sure what you meant. Can you provide me some examples? (like your own proposal on the shape)

pkong-ds replied:
How should I implement provider-specific fields?
Please find my attempt below, with some problems I ran into 🙏 

```golang
package config

var _ = Schema.Add("CaptchaConfig", `
{
	"type": "object",
	"additionalProperties": false,
	"properties": {
		"enabled": { "type": "boolean" },
		"providers": { "type": "array", "$ref": "#/$defs/CaptchaProvider" }
	}
}
`)

var _ = Schema.Add("CaptchaProvider", `
{
	"type": "string",
	"enum": ["cloudflare"]
}
`)

type CaptchaConfig struct {
	Enabled   bool                `json:"enabled,omitempty"`
	Providers []CaptchaProviderObject `json:"providers,omitempty"`
}

type CaptchaProviderObject interface {
	IsCaptchaProviderObject()
}

type CaptchaProviderType string

const (
	CaptchaProviderTypeCloudflare CaptchaProviderType = "cloudflare"
	CaptchaProviderTypeRecaptchaV2 CaptchaProviderType = "recaptcha_v2"
)

type CaptchaProviderAlias string

const (
	CaptchaProviderAliasCloudflare CaptchaProviderAlias = "cloudflare"
	CaptchaProviderAliasRecaptchaV2 CaptchaProviderAlias = "recaptcha_v2"
)

type CaptchaProvider struct {
	Type CaptchaProviderType `json:"type,omitempty"`
	Alias CaptchaProviderAlias `json:"alias,omitempty"`
}

func (p *CaptchaProvider) IsCaptchaProviderObject() {}
func (p *CaptchaProvider) GetType() CaptchaProviderType { return p.Type }
func (p *CaptchaProvider) GetAlias() CaptchaProviderAlias { return p.Alias }

var _ CaptchaProviderRoot = &CaptchaProvider{}

// Problem here, how do I define per-provider config?
// Maybe sth like below?

const (
	CaptchaProviderCloudflare = &CaptchaProvider{Type: CaptchaProviderTypeCloudflare, Alias: CaptchaProviderAliasCloudflare} // add cloudflare-specific fields here?
	CaptchaProviderRecaptchaV2 = &CaptchaProvider{Type: CaptchaProviderTypeRecaptchaV2, Alias: CaptchaProviderAliasRecaptchaV2} // add recaptcha-specific fields here?
)

// but above will throw &CaptchaProvider{…} (value of type *CaptchaProvider) is not constant
// what about var?

var (
	CaptchaProviderCloudflare = &CaptchaProvider{Type: CaptchaProviderTypeCloudflare, Alias: CaptchaProviderAliasCloudflare} // add cloudflare-specific fields here?
	CaptchaProviderRecaptchaV2 = &CaptchaProvider{Type: CaptchaProviderTypeRecaptchaV2, Alias: CaptchaProviderAliasRecaptchaV2} // add recaptcha-specific fields here?
)



```

pkong-ds replied:
Or maybe we can do it like AuthFlow, make these options as a property instead? Expected usage:

```
// authgear.yaml
captcha:
  cloudfare:
     secret: verysecret
  recaptchav2:
     secret: verysecret
```

pkong-ds replied:
This comment thread is resolved [here](https://github.com/authgear/authgear-server/pull/4337#discussion_r1639283104)

pkong-ds commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,354 @@
+
+Given `captcha.enabled=true` and `captcha.providers` is non-empty,
+
+1. The generated flow has `captcha.enabled=true` in its root configuration.
+2. All the branches of the first step (that is, the `identify` step) has `captcha.required=true`.
```
question: `root enabled=true` = `1st  required=true`?
| root `captcha.enabled` | 1st step `captcha.required`| 2nd step `captcha.required` | question: Which provider/step have captcha? |
|--------|--------|--------|--------|
| true | true | true | 1st & 2nd |
| false | false | false | none | 
| true | true | false | 1st only |
| true | false | false | 1st OR none? |
| false | true | false | 1st OR none? |
| false | false | true | 2nd OR none? |

louischan-oursky replied:
> root enabled=true = 1st  required=true?

Yes

---

| root `captcha.enabled` | Step N `captcha.required`| Step N+1 `captcha.required` | question: Which provider/step have captcha? |
|--------|--------|--------|--------|
| true | true | true | Step N and Step N+1. But by the time Step N+1 is reached, the captcha challenge is solved. The required flag has no actual effect. |
| false | false | false | none | 
| true | true | false | Step N |
| true | false | false | none |
| false | true | false | none |
| false | false | true | none |

pkong-ds commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,354 @@
+
+Captcha is supported only in the following step types:
+
+- `identify` in `signup`, `promote`, `login`, `signup_login`, and `account_recovery`.
```
Is `reauth` intentionally left out here? 🙏

louischan-oursky replied:
The `identify` step in `reauth` is done by an ID token, which involves no user interaction. So the captcha is supported in `authenticate` instead.

pkong-ds commented on docs/specs/authentication-flow-api-reference.md:
```diff
@@ -210,6 +231,14 @@ When you are in this step of this flow, you will see a response like the followi
+      "providers": [
+        {
+          "type": "cloudflare",
+          "alias": "cloudflare"
+        }
```
Do we need to show extra information like `site_key` here? 🙏

louischan-oursky replied:
No

pkong-ds commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,354 @@
+Given `captcha.enabled=true` and `captcha.providers` is non-empty,
+
+1. The generated flow has `captcha.enabled=true` in its root configuration.
+2. All the branches of the first step (that is, the `identify` step) has `captcha.required=true`.
```
> All the branches of the first step (that is, the `identify` step) has `captcha.required=true`.

> The identify step in reauth is done by an ID token, which involves no user interaction. So the captcha is supported in authenticate instead.

@louischan-oursky just to confirm, for `reauth` flow, `first step` here would refer to `authenticate`?

Context: adding captcha to `pkg/lib/authenticationflow/declarative/generate_config_reauth_flow.go`, wonder if I should add to `Identify` step or `Authenticate` step 🙏

pkong-ds commented on docs/specs/authentication-flow-api-reference.md:
```diff
@@ -239,6 +266,134 @@ When you are in this step of this flow, you will see a response like the followi
+  "risk_assessment": {
+    "alias": "recaptchav3",
+    "type": "recaptchav3",
+    "response": { ... }
```
nitpick: How about we use `response_token` for this field?

louischan-oursky replied:
Actually I do not know what `response` we will get from the client side library. My original intention is just pass what the client side library give us there.

pkong-ds replied:
> I do not know what response we will get from the client side library

Same, but since this field is captcha-provider-specific, I just thought it's safe to use `response_token` for `recaptchav3`, `recaptchav2` & `cloudflare`; We can always use another field name if some non-token response appear

But it's really trivial, can keep response if dun wanna change 🙏

louischan-oursky replied:
I use the name `response` because it is also the actual name the provider use in the API. The phrase "response token" only appears in the description.

pkong-ds commented on docs/specs/authentication-flow-api-reference.md:
```diff
@@ -239,6 +266,134 @@ When you are in this step of this flow, you will see a response like the followi
+
+#### risk_assessment input; type: recaptchav3
+
+- `risk_assessment.response`: The response provided by the reCAPTCHA v3 client-side library.
```
nitpick: How about we use `response_token` for this field?

> [response 	Required. The user response token provided by the reCAPTCHA client-side integration on your site.](https://developers.google.com/recaptcha/docs/verify#api_request)

louischan-oursky replied:
Do you mean renaming `response` to `response_token`? My original meaning is that `risk_assessment.response` is the `response` in `POST https://www.google.com/recaptcha/api/siteverify`.

pkong-ds replied:
> Do you mean renaming response to response_token?

yes, but rather trivial, just wanna be more descriptive to user

> My original meaning is that risk_assessment.response is the response in POST https://www.google.com/recaptcha/api/siteverify.

I thought it's whatever the client-side library returns 😱

pkong-ds commented on docs/specs/authentication-flow-api-reference.md:
```diff
@@ -239,6 +266,134 @@ When you are in this step of this flow, you will see a response like the followi
+  "captcha": {
+    "alias": "cloudflare",
+    "type": "cloudflare",
+    "response": { ... }
```
nitpick: What about we use `response_token` for this field?

pkong-ds commented on docs/specs/authentication-flow-api-reference.md:
```diff
@@ -239,6 +266,134 @@ When you are in this step of this flow, you will see a response like the followi
+
+#### Captcha input; type: cloudflare
+
+- `captcha.response`: The response provided by the Turnstile client-side library.
+
+#### Captcha input; type: recaptchav2
+
+- `captcha.response`: The response provided by the reCAPTCHA v2 client-side library.
```
nitpick: What about we use `response_token` for this field?

pkong-ds commented on docs/specs/authentication-flow-api-reference.md:
````diff
@@ -239,6 +266,134 @@ When you are in this step of this flow, you will see a response like the followi
+
+The shape of `risk_assessment`:
+
+```
````
```suggestion
```json
```

pkong-ds commented on docs/specs/authentication-flow-api-reference.md:
````diff
@@ -239,6 +266,134 @@ When you are in this step of this flow, you will see a response like the followi
+
+The shape of `captcha`:
+
+```
````
```suggestion
```json
```

pkong-ds commented on docs/specs/captcha.md:
````diff
@@ -1,41 +1,452 @@
+
+The configuration is as follows:
+
+```
````
```suggestion
```yaml
```

pkong-ds commented on docs/specs/captcha.md:
````diff
@@ -1,41 +1,452 @@
+
+For example,
+
+```
````
```suggestion
```yaml
```

pkong-ds commented on docs/specs/captcha.md:
````diff
@@ -1,41 +1,452 @@
+
+This can be achieved by customizing the flow.
+
+```
````
```suggestion
```yaml
```

pkong-ds commented on docs/specs/captcha.md:
````diff
@@ -1,41 +1,452 @@
+
+The developer can specify different Captcha provider to be used in different branches.
+
+```
````
```suggestion
```yaml
```

pkong-ds commented on docs/specs/captcha.md:
````diff
@@ -1,41 +1,452 @@
+
+Here is an example configuration:
+
+```
````
```suggestion
```yaml
```

pkong-ds commented on docs/specs/captcha.md:
````diff
@@ -1,41 +1,452 @@
+
+Here is an example configuration:
+
+```
````
```suggestion
```yaml
```

pkong-ds commented on docs/specs/captcha.md:
````diff
@@ -1,41 +1,452 @@
+
+Here is an example configuration:
+
+```
````
```suggestion
```yaml
```

pkong-ds commented on docs/specs/captcha_legacy.md:
````diff
@@ -0,0 +1,29 @@
+
+### authgear.yaml
+
+```
````
```suggestion
```yaml
```

pkong-ds reviewed (commented):
Also found some code blocks with syntax highlights are more readable [when I read in branch](https://github.com/louischan-oursky/authgear-server/blob/dev-1352-specify-captcha/docs/specs/captcha.md) 🙇

pkong-ds commented on docs/specs/captcha.md:
````diff
@@ -1,41 +1,452 @@
+    alias: cloudflare
+```
+
+- `captcha.mode`: When Captcha is required.
````
Missing risk_assessment field docs

pkong-ds commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,452 @@
+
+### Behavior of generated flows
+
+Given `captcha.enabled=true` and `captcha.providers` is non-empty,
+
+1. All the branches of the first step (that is, the `identify` step, or the `authenticate` step in reauth flow) has `captcha.mode=always`.
+2. The first provider in `captcha.providers` is used as `captcha.provider.alias`
+
+In terms of UX, when Captcha is enabled and configured, every generated flow requires captcha at the beginning of the flow.
```
Missing behavior of generated flows for risk_assessment

```md
Given `risk_assessment.enabled=true` and `risk_assessment.providers` is non-empty,
```

louischan-oursky replied:
OK. I will add it back.

pkong-ds commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,452 @@
+
+We can see that all supported step types have branches.
+
+To enable risk assessment and Captcha in a branch, add `risk_assessment` and `captcha` to the branch.
```
question: For a certain auth flow `flowA`, if one of `flowA`'s branch `branchX` enabled `captcha`, does below endpoints
- create - `POST /api/v1/authentication_flows`
- input - `POST /api/v1/authentication_flows/input`

i. always return `captcha` field; OR
ii. only return `captcha` field when the handler contains `branchX` logic?

```json
"result": {
        "state_token": "authflowstate_blahblah",
        "type": "signup", # "signup" | "login" | ...
        "name": "default",
         // below object
         "captcha": {
            "providers": [
                {
                    "type": "cloudflare",
                    "alias": "cloudflare"
                }
            ]
        },
```

louischan-oursky replied:
I have updated the api reference. We no longer return `captcha` in the root of the response. Instead, the captcha information is returned in the branch.

pkong-ds replied:
I see, thanks!

louischan-oursky commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,452 @@
+Type specific fields:
+
+- `risk_assessment.providers.type=recaptchav3.site_key`: The site key of reCAPTCHA v3.
+- `risk_assessment.providers.type=recaptchav3.risk_level`: The mapping of reCAPTCHA v3 score to Low, Medium, and High.
```
Yes. site_key and risk_level are both required.

pkong-ds replied:
Sor; I meant for _non-recaptchav3 providers_, should we make `risk_level` a required field too?

louischan-oursky replied:
OK. I will make it required for all providers.

chpapa commented on docs/specs/authentication-flow-api-reference.md:
```diff
@@ -219,7 +232,21 @@ When you are in this step of this flow, you will see a response like the followi
           {
-            "identification": "phone"
+            "identification": "phone",
+            "risk_assessment": {
```
`risk_assessment` sounds a bit too general? Since we're here specifically for detecting bot ("I'm not a robot" confirmation), what about something more specific to avoid future confusion with other kind of "risks"? e.g. impossible travel.

e.g. `bot_risk` or `bot_detection`

louischan-oursky replied:
What about something like

```yaml
captcha:
  enabled: true
  # This list contains the challenge-based providers
  providers:
  - type: cloudflare
  # risk_assessment is now nested under captcha. We can still have risk_assessment in the root in the future.
  risk_assessment:
    enabled: true
    providers:
    - type: recaptchav3
```

chpapa replied:
Sounds fair. (still... if we have a better name. Since "root" or not is not very easy for eye / still easy to confuse in documents, etc)

btw, why not...

```yaml
captcha:
  enabled: true
  # This list contains the challenge-based providers
  provider: cloudflare
  # risk_assessment is now nested under captcha. We can still have risk_assessment in the root in the future.
  risk_assessment:
    enabled: true
    provider: recaptchav3
```

louischan-oursky replied:
OK. let me revise to remove the array. So we only allow configuring one provider at a time.

chpapa commented on docs/specs/authentication-flow-api-reference.md:
```diff
@@ -219,7 +232,21 @@ When you are in this step of this flow, you will see a response like the followi
+            "identification": "phone",
+            "risk_assessment": {
+              "required": true,
+              "provider": {
```
Really a question: do we really need `type` and `alias`? In our use case probably impossible to support say two set of credentials of same type?

So we could keep it simple to just have one identifier for provider?

chpapa replied:
So something like....

```
"captcha": {
    "required": true,
    "provider": "cloudflare"
}

louischan-oursky replied:
I think in most case, we do not need to support two credentials of the same type. Let me remove `alias` for now. We can add it in the future if we need it.

chpapa commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,460 @@
 captcha:
-  provider: cloudflare
+  enabled: true
+  ip_allowlist:
```
does this also applies to `risk_assessment` automatically?

louischan-oursky replied:
Yes, it should.

chpapa replied:
maybe specify it or put the risk section under captcha as well.

louischan-oursky replied:
OK.

chpapa commented on docs/specs/captcha.md:
```diff
@@ -1,41 +1,460 @@
+---
+
+- `captcha.enabled`: If it is true, the new configuration is used.
+- `captcha.ip_allowlist`: A list of IPv4 CIDR notations or IPv4 addresses. If the incoming request matches any entry in the allowlist, the request bypasses Captcha.
```
IPv6?

louischan-oursky replied:
OK. Let me add it back.

chpapa reviewed (commented):
Did a very quick scan and leave some comments.

louischan-oursky commented:
@chpapa @pkong-ds I have updated the spec once again. The major changes are

1. Rename `captcha` to `bot_protection`
2. Place `risk_assessment` under `bot_protection`
3. `providers` becomes `provider`, that is, we support only one provider at a time.

Configuration example

```yaml
bot_protection:
  enabled: true
  provider:
    type: cloudflare
    site_key: "SITE_KEY"
  risk_assessment:
    enabled: true
    provider:
      type: recaptchav3
      site_key: "SITE_KEY"
      risk_score:
        low: 0.2
        medium: 0.5
        high: 0.7

authentication_flow:
  login_flows:
  - name: default
    steps:
    - type: identify
      one_of:
      - identification: email
    - type: authenticate
      one_of:
      - authentication: primary_password
        bot_protection:
          mode: "risk_level_high"
          provider:
            type: cloudflare
          risk_assessment:
            enabled: true
            provider:
              type: recaptchav3

```

pkong-ds commented:
@louischan-oursky seems I don't have access to push to `louischan-oursky/dev-1352-specify-captcha`. Submitted a PR to your branches instead LOL. ref https://github.com/louischan-oursky/authgear-server/pull/1

Please advise if otherwise~ 🙏

louischan-oursky commented:
@pkong-ds I added your commit to this PR. And also added a commit to update the audit log event name.

louischan-oursky commented:
@pkong-ds I added a commit based on your changes.

@chpapa @tung2744  @fungc-io  I added a commit to specify the configuration to support the behavior we discussed in the product meeting. https://github.com/authgear/authgear-server/pull/4282/commits/b8c90dcd7c056153ef7e821d812a53789d36d2aa#diff-7fa5fe90ad92ac79b786c50d705b525113599d205b1268f4d093a4249e8aeb03R60

pkong-ds commented on docs/specs/bot-protection.md:
```diff
@@ -122,9 +150,7 @@ Bot protection is supported only in the following step types:
-
-Note that the step type `create_authenticator` in `signup` and `promote` does not support bot protection.
-Bot protection should be configured at a previous step, for example `identify`.
+- `create_authenticator` in `signup` and `promote`
```
@louischan-oursky do you think we support all below

1. `create_authenticator_oob_otp`
2. `create_authenticator_password`
3. `create_authenticator_totp`

Or just 1? 🙏

louischan-oursky replied:
According to line 219, just Point 1 is supported.

pkong-ds commented on docs/specs/bot-protection.md:
```diff
@@ -210,16 +210,13 @@ The bot protection behavior in generated flows depend on
+|`bot_protection.authentication_flow.signup_or_login.mode`|Branches `identification: email`, `identification: phone`, `identification: username` of step `type: identify` in flows `signup`, `login`, `signup_login`, `promote` has `bot_protection.mode` equal to `bot_protection.authentication_flow.signup_or_login.mode`.|
+|`bot_protection.authentication_flow.account_recovery.mode`|All branches of `type: identify` of flow `account_recovery` has `bot_protection.mode` equal to `bot_protection.authentication_flow.account_recovery.mode`.|
+|`bot_protection.authenticator.password.mode`|Branches `authentication: primary_password`, `authentication: secondary_password` in step `type: authenticate` in flows `login` has `bot_protection.mode` equal to `bot_protection.authenticator.password.mode`.|
+|`bot_protection.authenticator.oob_otp_email.mode`|Branches `authentication: primary_oob_otp_email`, `authentication: secondary_oob_otp_email` in steps `type: authenticate`, `type: create_authenticator` in flows `signup`, `login`, `promote` has `bot_protection.mode` equal to `bot_protection.authenticator.oob_otp_email.mode`.|
```
Seems 

- flows `signup`, `promote`
     - branches `identification: email` `identification: phone` 
         - `type: identify` 

should equal to `bot_protection.authenticator.oob_otp_email.mode`

## Rationale

Currently AuthUI will send sms/email immediately after identify with `phone/email`.

### Screenshot Video

https://github.com/user-attachments/assets/466455f9-1386-4960-888b-180d541d58ce

louischan-oursky replied:
Added commit "Apply bot protection to signup verification as well" to clarify this case.

pkong-ds commented on docs/specs/bot-protection.md:
```diff
@@ -0,0 +1,448 @@
+|`bot_protection.authentication_flow.account_recovery.mode`|All branches of `type: identify` of flow `account_recovery` has `bot_protection.mode` equal to `bot_protection.authentication_flow.account_recovery.mode`.|
+|`bot_protection.authenticator.password.mode`|Branches `authentication: primary_password`, `authentication: secondary_password` in step `type: authenticate` in flows `login` has `bot_protection.mode` equal to `bot_protection.authenticator.password.mode`.|
+|`bot_protection.authenticator.oob_otp_email.mode`|Branches `authentication: primary_oob_otp_email`, `authentication: secondary_oob_otp_email` in steps `type: authenticate`, `type: create_authenticator` in flows `signup`, `login`, `promote` has `bot_protection.mode` equal to `bot_protection.authenticator.oob_otp_email.mode`.|
+|`verification.claims.email.required`|If `verification.claims.email.required` is true, then branches `identification: email` in step `type: identify` in flows `signup`, `promote` has `bot_protection.mode` equal to `bot_protection.authenticator.oob_otp_email.mode`.|
```
Seems this conflicts with `bot_protection.authentication_flow.signup_or_login.mode`

What should

> branch `identification: email` in step `type: identify` in flows `signup`, `promote` 

be, in case of below config?

```
bot_protection.authentication_flow.signup_or_login.mode: never
bot_protection.authenticator.oob_otp_email.mode: always
verification.claims.email.required: true
```

Maybe add below?

```md
#### Behavior of generated flow in case of conflicts in configuration
In case of any conflicts, the stricter [Risk Mode](#risk-level-mode) will be taken, where 

     `always` > `risk_level_high` > `risk_level_medium` > `never` 

in terms of strictness
```

louischan-oursky replied:
Yes, you're correct.

louischan-oursky replied:
Added commit "eClarify what mode is used when there are multiple contributions"

pkong-ds commented on docs/specs/event.md:
```diff
@@ -208,6 +211,7 @@ Use this event to add custom fields to the JWT access token.
 - [authentication.secondary.oob_otp_email.failed](#authenticationsecondaryoob-otp-emailfailed)
 - [authentication.secondary.oob_otp_sms.failed](#authenticationsecondaryoob-otp-smsfailed)
 - [authentication.secondary.recovery_code.failed](#authenticationsecondaryrecovery-codefailed)
+- [captcha.failed](#captchafailed)
```
[bot_protection.verification.failed](#bot-protectionverificationfailed)

pkong-ds commented on docs/specs/bot-protection.md:
```diff
@@ -0,0 +1,450 @@
+|`bot_protection.authentication_flow.account_recovery.mode`|Contribute to all branches of `type: identify` of flow `account_recovery` has `bot_protection.mode` equal to `bot_protection.authentication_flow.account_recovery.mode`.|
+|`bot_protection.authenticator.password.mode`|Contribute to branches `authentication: primary_password`, `authentication: secondary_password` in step `type: authenticate` in flows `login`.|
+|`bot_protection.authenticator.oob_otp_email.mode`|Contribute to branches `authentication: primary_oob_otp_email`, `authentication: secondary_oob_otp_email` in steps `type: authenticate`, `type: create_authenticator` in flows `signup`, `login`, `promote`.|
+|`verification.claims.email.required`|If `verification.claims.email.required` is true, contribute to branches `identification: email` in step `type: identify` in flows `signup`, `promote`.|
```
On 2nd thoughts, 

is there case where
1. `signup identify by email`, but `email` is NOT sent in next immediate step?
2. `signup identify by phone`, but `phone code` is NOT sent in next immediate step?

I tried both `w/` & `w/o` verify, seems an email/phone code are always sent in next immediate step, w/o extra user interaction

Therefore I suggest:

for `bot_protection.authentication_flow.authenticator.oob_otp_email.mode`  (& sms), it should contribute to

1. ALWAYS `identification: email` in step `type: identify` in flows `signup`, `promote`
2. `authentication: primary_oob_otp_email`, `authentication: secondary_oob_otp_email` in steps `type: authenticate` in flows `login`
3. `authentication: secondary_oob_otp_email` in steps `type: create_authenticator` in flows `signup`, `promote`

Note how `primary_oob_otp_email` is ignored in 3, because it is handled by 1 already

Similarly for `sms`

louischan-oursky replied:
> signup identify by email, but email is NOT sent in next immediate step?

Yes, Use password to sign up, and turn off verification.

pkong-ds replied:
I tried with below custom flow, to no avail.

Error is `invalid value` in authUI enter password page

```yaml
authentication_flow:
  signup_flows:
  - name: default
    steps:
    - name: signup_identify
      type: identify
      one_of:
      - identification: email
        steps:
        - name: authenticate_primary_password
          type: create_authenticator
          one_of:
          - authentication: primary_password
        - name: authenticate_primary_oob_otp_email
          type: create_authenticator
          one_of:
          - authentication: primary_oob_otp_email
```

### base64-decoded error
```
{
  "Form": {
    "gorilla.csrf.Token": [
      "FqOqdERU8uGw1l5ZBY/7syLlIxbDcDVQACY0faDmcIF+y66C3zHKWZwXLxIMXxhto6THBs+pmGQC1hGMtUvX/A=="
    ],
    "q_login_id": ["abc@oursky.com"],
    "q_login_id_input_type": ["text"],
    "q_login_id_key": ["email"],
    "q_login_id_type": ["email"],
    "x_action": ["login_id"]
  },
  "Error": {
    "name": "Invalid",
    "reason": "ValidationFailed",
    "message": "invalid value",
    "code": 400,
    "info": {
      "causes": [
        {
          "location": "",
          "kind": "required",
          "details": {
            "actual": ["authentication", "channel"],
            "expected": ["authentication", "target"],
            "missing": ["target"]
          }
        }
      ]
    }
  }
```

pkong-ds commented:
@louischan-oursky added commit for new agreed spec 🙏 https://github.com/authgear/authgear-server/pull/4282/commits/dec4edc15245a63f3f1cc11013959132cf97a96f

## Preview

<img width="530" alt="image" src="https://github.com/user-attachments/assets/ccd01100-fabd-4c34-a152-7bcf5e3798e4">

pkong-ds commented:
@louischan-oursky added 2 commits to for product meeting discussion result

pkong-ds commented:
Added [Remove all verify-related spec](https://github.com/authgear/authgear-server/pull/4282/commits/dde3a2621a30d65dd40b726fb5e541faf830f699)

louischan-oursky commented:
@tung2744 This should be good to merge.
